experiment01-tailgrow/xdp_prog_kern2.c

```diff
@@ -0,0 +1,63 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#include <linux/bpf.h>
+#include <bpf/bpf_helpers.h>
+
+#define MTU 1536
+#define MIN_LEN 14
+
+/*
+ * This example show howto access packet last byte in XDP packet,
+ * without parsing packet contents.
+ *
+ * It is not very effecient, as it advance the data pointer one-byte in a
+ * loop until reaching data_end.  This is needed as the verifier only allows
+ * accessing data via advancing the position of the data pointer. The bounded
+ * loop with a max number of iterations allows the verifier to see the bound.
+ */
+
+SEC("xdp_end_loop")
+int _xdp_end_loop(struct xdp_md *ctx)
+{
+	void *data_end = (void *)(long)ctx->data_end;
+	void *data     = (void *)(long)ctx->data;
+	unsigned char *ptr;
+	unsigned int i;
+	void *pos;
+
+	/* Assume minimum length to reduce loops needed a bit */
+	unsigned int offset = MIN_LEN;
```

Comment:
If you're going to optimise, couldn't you technically make the loop a binary search instead? :)

Reply:
Yes, to binary-search, but I wonder/worry how the BPF-verifiers bounded-loops verification would handle that ;-)